{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "392dfcd0-53a8-4779-b74a-b791f63acedb",
   "metadata": {},
   "source": [
    "# Homework 2: Homework 2: Intro to Python programming for data analysis\n",
    "\n",
    "Instructions on the [course webpage](https://bengtssonpalme.github.io/MPBIO-BBT045-2024/python_for_data_analysis/homework_python_intro.html)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fef8ae62-a9b6-4f76-b2ef-c17ee2a9bd59",
   "metadata": {},
   "source": [
    "## Setup\n",
    "Files needed in the same directory as this notebook:\n",
    "- S288C_reference_sequence_R64-4-1_20230830.fsa\n",
    "- saccharomyces_cerevisiae_R64-4-1_20230830.gff\n",
    "- Y55_JRIF00000000_SGD_cds.fsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a734d8e-51a9-4fdf-8a44-228c846ec14a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/tmp.5733315/ipykernel_1096114/1453496124.py:2: DeprecationWarning: \n",
      "Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),\n",
      "(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)\n",
      "but was not found to be installed on your system.\n",
      "If this would cause problems for you,\n",
      "please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466\n",
      "        \n",
      "  import pandas as pd # enables manipulation of dataframes in Python\n"
     ]
    }
   ],
   "source": [
    "# load modules\n",
    "import pandas as pd # enables manipulation of dataframes in Python\n",
    "from Bio import SeqIO # enables manipulation of sequence files\n",
    "from Bio.SeqUtils import GC # calculate GC content with Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37668892-3f26-468d-a8f1-823b379918af",
   "metadata": {},
   "source": [
    "## Task 1: How many genes are on yeast S288C chromosome II?\n",
    "Read in the file saccharomyces_cerevisiae_R64-4-1_20230830.gff.\n",
    "\n",
    "Count the number of genes on chromosome II using dataframe filtering and manipulation functions like you've seen in the tutorial. Please have your code produce a numerical result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "203a72ec-b8af-4163-b096-1d8377684eec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "456"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the filename and the column names\n",
    "filename = \"saccharomyces_cerevisiae_R64-4-1_20230830.gff\"\n",
    "col_names = [\"chromosome\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"]\n",
    "\n",
    "# Lenght of the FASTA at the end of the file\n",
    "length_FASTA = 180335-28343\n",
    "\n",
    "# Read in the gff file, skipping the first 21 rows (info) and the last 151992 rows (FASTA)\n",
    "saccharomyces_cerevisiae_df = pd.read_csv(filename, sep = '\\t', names = col_names, skiprows = 21, skipfooter = length_FASTA, engine = 'python')\n",
    "\n",
    "# Filter every line which has chromosome set to chrII\n",
    "ChrII = saccharomyces_cerevisiae_df.loc[saccharomyces_cerevisiae_df.chromosome == \"chrII\"]\n",
    "# Filter for genes\n",
    "ChrII_genes = ChrII.loc[ChrII.feature == \"gene\"]\n",
    "\n",
    "# Display the number of rows in the dataframe\n",
    "len(ChrII_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76004fcc-7c30-4e00-92c9-8098d3f31ad8",
   "metadata": {},
   "source": [
    "## Task 2: Compute average gene length by chromosome in the S288C yeast strain\n",
    "Using the same dataframe you used in the Task 1, calculate the average gene length per chromosome. Remember to inspect your data, to see what you're working with. Please use the Pandas dataframe manipulation methods we've covered in the tutorial, as much as you can."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "454a2784-ec94-4bdb-acab-f0d936353a7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chromosome\n",
       "chrI       1257.40\n",
       "chrII      1336.69\n",
       "chrIII     1203.89\n",
       "chrIV      1341.95\n",
       "chrIX      1337.31\n",
       "chrV       1237.86\n",
       "chrVI      1330.88\n",
       "chrVII     1342.01\n",
       "chrVIII    1268.29\n",
       "chrX       1398.16\n",
       "chrXI      1420.87\n",
       "chrXII     1369.62\n",
       "chrXIII    1370.18\n",
       "chrXIV     1357.89\n",
       "chrXV      1318.05\n",
       "chrXVI     1347.07\n",
       "chrmt      2572.43\n",
       "Name: length, dtype: float64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for genes\n",
    "genes_Saccharomyces = saccharomyces_cerevisiae_df.loc[saccharomyces_cerevisiae_df.feature == \"gene\"]\n",
    "\n",
    "# Create a column with the length of the gene\n",
    "genes_Saccharomyces = genes_Saccharomyces.assign(length=lambda df: df.end-df.start)\n",
    "# Source `assign`and `lambda` function: pandas cheat sheet https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf\n",
    "\n",
    "# Group by chromosome\n",
    "grouped_genes_chromosome = genes_Saccharomyces.groupby('chromosome')\n",
    "\n",
    "# Display the average of the length for each group (chromosome) using `mean` and `round`\n",
    "round(grouped_genes_chromosome.length.mean(), 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f590fc4f-3b1f-4160-87e2-f565a0b4f4b1",
   "metadata": {},
   "source": [
    "## Task 3: Extract gene IDs from the S288C yeast GFF file and look for some genes\n",
    "\n",
    "Here's a function that extracts the gene ID from the column \"attributes\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "712b8b44-4043-4608-a567-d793ac796192",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gene_id_from_attributes_s288c(df_name): \n",
    "    # this function requires the argument of the GFF dataframe name be given to it\n",
    "    # it retrieves the IDs in a Pandas series from the attributes column of the GFF file\n",
    "    # first save the dataframe to a temporary version to work with\n",
    "    tmp_df = df_name.copy()\n",
    "    # split the contents of the attributes column into a list using the semicolons as delimiters\n",
    "    tmp_series = tmp_df.attributes.str.split(\";\")\n",
    "    # overwrite the contents of the attributes column into lists instead of strings \n",
    "    tmp_df[\"attributes\"] = tmp_series\n",
    "    # extract the first element of each list into a new series\n",
    "    tmp_series = tmp_df.attributes.str[0]\n",
    "    # split that item into a new list and save only the second element to the series\n",
    "    tmp_series = tmp_series.str.split(\"=\").str[1]\n",
    "    # the series containing the gene IDs will be returned\n",
    "    return tmp_series"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a9c3817-b847-44f6-9c88-58951ebdfee1",
   "metadata": {},
   "source": [
    "a. Using the table manipulation functions, create a new column gene_id in the GFF table. You should find clues for how to do this in the last exercise of the tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a085ba47-e7ba-451a-b580-7197e915ca50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>230218</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chrI;dbxref=NCBI:BK006935.2;Name=chrI</td>\n",
       "      <td>chrI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>telomere</td>\n",
       "      <td>1</td>\n",
       "      <td>801</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=TEL01L;Name=TEL01L;Note=Telomeric%20region%...</td>\n",
       "      <td>TEL01L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>X_element</td>\n",
       "      <td>337</td>\n",
       "      <td>801</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=TEL01L;Name=TEL01L_X_element</td>\n",
       "      <td>TEL01L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>X_element_combinatorial_repeat</td>\n",
       "      <td>63</td>\n",
       "      <td>336</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=TEL01L;Name=TEL01L_X_element_combinator...</td>\n",
       "      <td>TEL01L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>telomeric_repeat</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=TEL01L;Name=TEL01L_telomeric_repeat_1</td>\n",
       "      <td>TEL01L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28317</th>\n",
       "      <td>chrmt</td>\n",
       "      <td>SGD</td>\n",
       "      <td>noncoding_exon</td>\n",
       "      <td>85295</td>\n",
       "      <td>85777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=YNCQ0027W_ncRNA;Name=YNCQ0027W_noncodin...</td>\n",
       "      <td>YNCQ0027W_ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28318</th>\n",
       "      <td>chrmt</td>\n",
       "      <td>SGD</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>85295</td>\n",
       "      <td>85777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YNCQ0027W_ncRNA;Name=YNCQ0027W_ncRNA;Parent...</td>\n",
       "      <td>YNCQ0027W_ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28319</th>\n",
       "      <td>chrmt</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>85554</td>\n",
       "      <td>85709</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Q0297;Name=Q0297;Alias=ORF12;Ontology_term=...</td>\n",
       "      <td>Q0297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28320</th>\n",
       "      <td>chrmt</td>\n",
       "      <td>SGD</td>\n",
       "      <td>CDS</td>\n",
       "      <td>85554</td>\n",
       "      <td>85709</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=Q0297_mRNA;Name=Q0297_CDS;orf_classific...</td>\n",
       "      <td>Q0297_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28321</th>\n",
       "      <td>chrmt</td>\n",
       "      <td>SGD</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>85554</td>\n",
       "      <td>85709</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Q0297_mRNA;Name=Q0297_mRNA;Parent=Q0297</td>\n",
       "      <td>Q0297_mRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28322 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome source                         feature  start     end score  \\\n",
       "0           chrI    SGD                      chromosome      1  230218     .   \n",
       "1           chrI    SGD                        telomere      1     801     .   \n",
       "2           chrI    SGD                       X_element    337     801     .   \n",
       "3           chrI    SGD  X_element_combinatorial_repeat     63     336     .   \n",
       "4           chrI    SGD                telomeric_repeat      1      62     .   \n",
       "...          ...    ...                             ...    ...     ...   ...   \n",
       "28317      chrmt    SGD                  noncoding_exon  85295   85777     .   \n",
       "28318      chrmt    SGD                           ncRNA  85295   85777     .   \n",
       "28319      chrmt    SGD                            gene  85554   85709     .   \n",
       "28320      chrmt    SGD                             CDS  85554   85709     .   \n",
       "28321      chrmt    SGD                            mRNA  85554   85709     .   \n",
       "\n",
       "      strand phase                                         attributes  \\\n",
       "0          .     .           ID=chrI;dbxref=NCBI:BK006935.2;Name=chrI   \n",
       "1          -     .  ID=TEL01L;Name=TEL01L;Note=Telomeric%20region%...   \n",
       "2          -     .                Parent=TEL01L;Name=TEL01L_X_element   \n",
       "3          -     .  Parent=TEL01L;Name=TEL01L_X_element_combinator...   \n",
       "4          -     .       Parent=TEL01L;Name=TEL01L_telomeric_repeat_1   \n",
       "...      ...   ...                                                ...   \n",
       "28317      +     .  Parent=YNCQ0027W_ncRNA;Name=YNCQ0027W_noncodin...   \n",
       "28318      +     .  ID=YNCQ0027W_ncRNA;Name=YNCQ0027W_ncRNA;Parent...   \n",
       "28319      +     .  ID=Q0297;Name=Q0297;Alias=ORF12;Ontology_term=...   \n",
       "28320      +     0  Parent=Q0297_mRNA;Name=Q0297_CDS;orf_classific...   \n",
       "28321      +     .         ID=Q0297_mRNA;Name=Q0297_mRNA;Parent=Q0297   \n",
       "\n",
       "               gene_ID  \n",
       "0                 chrI  \n",
       "1               TEL01L  \n",
       "2               TEL01L  \n",
       "3               TEL01L  \n",
       "4               TEL01L  \n",
       "...                ...  \n",
       "28317  YNCQ0027W_ncRNA  \n",
       "28318  YNCQ0027W_ncRNA  \n",
       "28319            Q0297  \n",
       "28320       Q0297_mRNA  \n",
       "28321       Q0297_mRNA  \n",
       "\n",
       "[28322 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Feels weird to do this to everything and not just the genes, but here is the code for what is asked ¯\\_(ツ)_/¯\n",
    "\n",
    "# Extract the gene IDs using the function provided\n",
    "gene_IDs = extract_gene_id_from_attributes_s288c(saccharomyces_cerevisiae_df)\n",
    "\n",
    "# Create a new column called gene_ID and assign the values extracted above to it\n",
    "saccharomyces_cerevisiae_df['gene_ID'] = gene_IDs\n",
    "\n",
    "# Display the result (may need to scroll to the right to see the column)\n",
    "saccharomyces_cerevisiae_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "900d867a-20d3-44cd-94b3-55caf6db3d1d",
   "metadata": {},
   "source": [
    "b. Write code (using the table manipulation functions) that \n",
    "1) checks whether the following gene IDs are present in the table and\n",
    "2) only prints the chromosomes on which these are found.\n",
    "\n",
    "Hint: For 1), check out the isin() function though it's not the only way.\n",
    "- YJR117W\n",
    "- YKL156C-A\n",
    "- YOL110W\n",
    "- YJR104C\n",
    "- YJL111W\n",
    "- YXYZZY\n",
    "- YNOPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a93b188d-9aa7-4bb3-8f70-989d832fda56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14282</th>\n",
       "      <td>chrX</td>\n",
       "      <td>YJL111W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15221</th>\n",
       "      <td>chrX</td>\n",
       "      <td>YJR104C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278</th>\n",
       "      <td>chrX</td>\n",
       "      <td>YJR117W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15770</th>\n",
       "      <td>chrXI</td>\n",
       "      <td>YKL156C-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23662</th>\n",
       "      <td>chrXV</td>\n",
       "      <td>YOL110W</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome    gene_ID\n",
       "14282       chrX    YJL111W\n",
       "15221       chrX    YJR104C\n",
       "15278       chrX    YJR117W\n",
       "15770      chrXI  YKL156C-A\n",
       "23662      chrXV    YOL110W"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a list of gene IDs to search for\n",
    "gene_ID_search_list = [\"YJR117W\", \"YKL156C-A\", \"YOL110W\", \"YJR104C\", \"YJL111W\", \"YXYZZY\", \"YNOPE\"]\n",
    "\n",
    "# Use `isin()` to search for the values in this list in the gene_ID column created above:\n",
    "found_geneIDs = saccharomyces_cerevisiae_df[saccharomyces_cerevisiae_df.gene_ID.isin(gene_ID_search_list)]\n",
    "\n",
    "# Display the columns \"chromsome\" and \"gene_ID\"\n",
    "found_geneIDs[[\"chromosome\", \"gene_ID\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "884cc5d6-565d-45ae-935a-dcdc4482fae0",
   "metadata": {},
   "source": [
    "## Task 4: Find top 5 longest genes in the S288C yeast\n",
    "Compute gene lengths as before, but this time simply get the highest 5 values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ce368df3-ce0c-4110-9415-a793e91777c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17649</th>\n",
       "      <td>chrXII</td>\n",
       "      <td>YLR106C</td>\n",
       "      <td>14732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28138</th>\n",
       "      <td>chrmt</td>\n",
       "      <td>Q0045</td>\n",
       "      <td>12883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16693</th>\n",
       "      <td>chrXI</td>\n",
       "      <td>YKR054C</td>\n",
       "      <td>12278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12176</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>YHR099W</td>\n",
       "      <td>11234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6385</th>\n",
       "      <td>chrIV</td>\n",
       "      <td>YDR457W</td>\n",
       "      <td>9806</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome  gene_ID  length\n",
       "17649     chrXII  YLR106C   14732\n",
       "28138      chrmt    Q0045   12883\n",
       "16693      chrXI  YKR054C   12278\n",
       "12176    chrVIII  YHR099W   11234\n",
       "6385       chrIV  YDR457W    9806"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From task 2: \n",
    "# Filter for genes\n",
    "genes_Saccharomyces = saccharomyces_cerevisiae_df.loc[saccharomyces_cerevisiae_df.feature == \"gene\"]\n",
    "\n",
    "# Create a column with the length of the gene\n",
    "genes_Saccharomyces = genes_Saccharomyces.assign(length=lambda df: df.end-df.start)\n",
    "# Source `assign`and `lambda`: pandas cheat sheet https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf\n",
    "\n",
    "\n",
    "# New:\n",
    "# Display the n longest (in this case 5) genes, and only display three columns (chr, ID, length)\n",
    "genes_Saccharomyces.nlargest(5, 'length')[[\"chromosome\", \"gene_ID\", \"length\"]]\n",
    "# Source `nlargest`: pandas cheat sheet https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "198586f0-e853-47ef-8ef7-f851a4f88d02",
   "metadata": {},
   "source": [
    "## Task 5: Calculations per gene in the Y55 strain"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23618dbd-5795-4995-b9f7-a2b28af4cd77",
   "metadata": {},
   "source": [
    "Convert the FASTA file to a dataframe using the DataFrame() function, setting the columns as `gene_info` and sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98844cd6-388e-47e0-a8de-b643b72fc536",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%script false --no-raise-error\n",
    "\n",
    "# create a Pandas dataframe from the FASTA file\n",
    "# ref: https://stackoverflow.com/questions/19436789/biopython-seqio-to-pandas-dataframe\n",
    "# ref: https://stackoverflow.com/questions/69696077/i-want-to-parse-sequences-and-sequence-ids-from-a-fasta-file-and-assign-them-to\n",
    "\n",
    "input_fasta = \"Y55_JRIF00000000_SGD_cds.fsa\"\n",
    "\n",
    "# create two empty lists for the sequence IDs and sequences\n",
    "id_list = []\n",
    "sequence_list = []\n",
    "\n",
    "with open(input_fasta) as fasta_file: \n",
    "    # open the file\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        # parse the contents of the FASTA file\n",
    "        # and append the relevant information to the lists\n",
    "        id_list.append(seq_record.description)\n",
    "        sequence_list.append(str(seq_record.seq))\n",
    "\n",
    "# generate a new Pandas dataframe from the two lists\n",
    "# ref: https://www.geeksforgeeks.org/create-a-pandas-dataframe-from-lists/\n",
    "genes_y55_df = pd.DataFrame(list(zip(id_list, sequence_list)), columns = ['gene_info', 'sequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24417d9-5ba4-4f0a-a9ee-c2efcb082e78",
   "metadata": {},
   "source": [
    "Provided function to extract the gene ID from the `gene_info` fields:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "41106af8-bb97-4823-8d5f-e5e95c7f1436",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to extract the gene IDs from the gene_info columns\n",
    "\n",
    "def extract_gene_id_from_names_y55(gene_info_col):\n",
    "    # define the function name & set internal argument\n",
    "    return gene_info_col.split(\"|\")[3]\n",
    "    # split the string into a list using pipes (\"|\") as the delimiters\n",
    "    # then take the 4th element in the list (Pythonic index 3)\n",
    "    # return that value to outside of the function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac40bc2-bb40-4eee-a58c-73ebede96883",
   "metadata": {},
   "source": [
    "a. Use the function to add a gene_id column.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6ee8f92c-d5b6-4e23-bc9c-da658310d3f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1-10...</td>\n",
       "      <td>ATGAAGTTGAACATTTCTTACCCAGTCAACGGGTCTCAAAAGACCT...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1794...</td>\n",
       "      <td>ATGGGTAGAAGAAAAATTGAAATTGAACCTATCAAAGATGATAGAA...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 4404...</td>\n",
       "      <td>ATGGGAATATTTCGTTGGAACTATCCAGAGAGTTCTGTCCCCGGCG...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 5652...</td>\n",
       "      <td>ATGTACCAAAATAATGTATTGAATGCTATTTTGGCTTCTGAAAAAT...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 7411...</td>\n",
       "      <td>ATGAGTGTATTAAGATCTACATGCCTTTTCTTCCCTCCAAGATCCT...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5325</th>\n",
       "      <td>YNL335W_Y55 gi|696433996|gb|JRIF01000367.1| 55...</td>\n",
       "      <td>ATGTCACAGTACGGATTTGTAAGAGTTCCTAGAGAGGTAGAAAAGG...</td>\n",
       "      <td>JRIF01000367.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5326</th>\n",
       "      <td>YNL334C_Y55 SNO2 gi|696433996|gb|JRIF01000367....</td>\n",
       "      <td>ATGACAGTAAAGGATAAAAATCAACTAGCTCGATGTGATGCATTGA...</td>\n",
       "      <td>JRIF01000367.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5327</th>\n",
       "      <td>UNDEF_Y55 gi|696433907|gb|JRIF01000396.1| 1501...</td>\n",
       "      <td>ATGGACGACATTGAAACAGCCAAGAATCTGACGGTAAAAGCACGTA...</td>\n",
       "      <td>JRIF01000396.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5328</th>\n",
       "      <td>UNDEF_Y55 gi|696433883|gb|JRIF01000403.1| 1372...</td>\n",
       "      <td>ATGCCTTATAAAACAGCTATAGATTGCATAGAAGAGTTAGCTACTC...</td>\n",
       "      <td>JRIF01000403.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5329</th>\n",
       "      <td>UNDEF_Y55 gi|696433878|gb|JRIF01000404.1| 1256...</td>\n",
       "      <td>ATGTTAAGACAAATCCAAGATAAATTTGGTGGTTCTGTAAAATTAA...</td>\n",
       "      <td>JRIF01000404.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5330 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              gene_info  \\\n",
       "0     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1-10...   \n",
       "1     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1794...   \n",
       "2     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 4404...   \n",
       "3     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 5652...   \n",
       "4     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 7411...   \n",
       "...                                                 ...   \n",
       "5325  YNL335W_Y55 gi|696433996|gb|JRIF01000367.1| 55...   \n",
       "5326  YNL334C_Y55 SNO2 gi|696433996|gb|JRIF01000367....   \n",
       "5327  UNDEF_Y55 gi|696433907|gb|JRIF01000396.1| 1501...   \n",
       "5328  UNDEF_Y55 gi|696433883|gb|JRIF01000403.1| 1372...   \n",
       "5329  UNDEF_Y55 gi|696433878|gb|JRIF01000404.1| 1256...   \n",
       "\n",
       "                                               sequence         gene_ID  \n",
       "0     ATGAAGTTGAACATTTCTTACCCAGTCAACGGGTCTCAAAAGACCT...  JRIF01000001.1  \n",
       "1     ATGGGTAGAAGAAAAATTGAAATTGAACCTATCAAAGATGATAGAA...  JRIF01000001.1  \n",
       "2     ATGGGAATATTTCGTTGGAACTATCCAGAGAGTTCTGTCCCCGGCG...  JRIF01000001.1  \n",
       "3     ATGTACCAAAATAATGTATTGAATGCTATTTTGGCTTCTGAAAAAT...  JRIF01000001.1  \n",
       "4     ATGAGTGTATTAAGATCTACATGCCTTTTCTTCCCTCCAAGATCCT...  JRIF01000001.1  \n",
       "...                                                 ...             ...  \n",
       "5325  ATGTCACAGTACGGATTTGTAAGAGTTCCTAGAGAGGTAGAAAAGG...  JRIF01000367.1  \n",
       "5326  ATGACAGTAAAGGATAAAAATCAACTAGCTCGATGTGATGCATTGA...  JRIF01000367.1  \n",
       "5327  ATGGACGACATTGAAACAGCCAAGAATCTGACGGTAAAAGCACGTA...  JRIF01000396.1  \n",
       "5328  ATGCCTTATAAAACAGCTATAGATTGCATAGAAGAGTTAGCTACTC...  JRIF01000403.1  \n",
       "5329  ATGTTAAGACAAATCCAAGATAAATTTGGTGGTTCTGTAAAATTAA...  JRIF01000404.1  \n",
       "\n",
       "[5330 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply the function provided above to the column gene_info\n",
    "gene_IDs_Y55 = genes_y55_df.gene_info.apply(extract_gene_id_from_names_y55)\n",
    "\n",
    "# Add the gene IDs as a new column\n",
    "genes_y55_df['gene_ID'] = gene_IDs_Y55\n",
    "\n",
    "# Display it\n",
    "genes_y55_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35826a4a-3150-40a4-9c4e-9f2b11199029",
   "metadata": {},
   "source": [
    "b. Rare codons are biologically significant. In yeast, the rarest codon (besides stop codons) is CGG (mRNA). So, using dataframe manipulation functions on the resulting dataframe from a., create a column to count the occurrence of GCC (DNA) per gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c2f21efa-74b3-4e4d-b0a1-fe62c3e3e9d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>count_rare_codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1-10...</td>\n",
       "      <td>ATGAAGTTGAACATTTCTTACCCAGTCAACGGGTCTCAAAAGACCT...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1794...</td>\n",
       "      <td>ATGGGTAGAAGAAAAATTGAAATTGAACCTATCAAAGATGATAGAA...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 4404...</td>\n",
       "      <td>ATGGGAATATTTCGTTGGAACTATCCAGAGAGTTCTGTCCCCGGCG...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 5652...</td>\n",
       "      <td>ATGTACCAAAATAATGTATTGAATGCTATTTTGGCTTCTGAAAAAT...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 7411...</td>\n",
       "      <td>ATGAGTGTATTAAGATCTACATGCCTTTTCTTCCCTCCAAGATCCT...</td>\n",
       "      <td>JRIF01000001.1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5325</th>\n",
       "      <td>YNL335W_Y55 gi|696433996|gb|JRIF01000367.1| 55...</td>\n",
       "      <td>ATGTCACAGTACGGATTTGTAAGAGTTCCTAGAGAGGTAGAAAAGG...</td>\n",
       "      <td>JRIF01000367.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5326</th>\n",
       "      <td>YNL334C_Y55 SNO2 gi|696433996|gb|JRIF01000367....</td>\n",
       "      <td>ATGACAGTAAAGGATAAAAATCAACTAGCTCGATGTGATGCATTGA...</td>\n",
       "      <td>JRIF01000367.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5327</th>\n",
       "      <td>UNDEF_Y55 gi|696433907|gb|JRIF01000396.1| 1501...</td>\n",
       "      <td>ATGGACGACATTGAAACAGCCAAGAATCTGACGGTAAAAGCACGTA...</td>\n",
       "      <td>JRIF01000396.1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5328</th>\n",
       "      <td>UNDEF_Y55 gi|696433883|gb|JRIF01000403.1| 1372...</td>\n",
       "      <td>ATGCCTTATAAAACAGCTATAGATTGCATAGAAGAGTTAGCTACTC...</td>\n",
       "      <td>JRIF01000403.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5329</th>\n",
       "      <td>UNDEF_Y55 gi|696433878|gb|JRIF01000404.1| 1256...</td>\n",
       "      <td>ATGTTAAGACAAATCCAAGATAAATTTGGTGGTTCTGTAAAATTAA...</td>\n",
       "      <td>JRIF01000404.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5330 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              gene_info  \\\n",
       "0     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1-10...   \n",
       "1     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 1794...   \n",
       "2     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 4404...   \n",
       "3     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 5652...   \n",
       "4     UNDEF_Y55 gi|696435221|gb|JRIF01000001.1| 7411...   \n",
       "...                                                 ...   \n",
       "5325  YNL335W_Y55 gi|696433996|gb|JRIF01000367.1| 55...   \n",
       "5326  YNL334C_Y55 SNO2 gi|696433996|gb|JRIF01000367....   \n",
       "5327  UNDEF_Y55 gi|696433907|gb|JRIF01000396.1| 1501...   \n",
       "5328  UNDEF_Y55 gi|696433883|gb|JRIF01000403.1| 1372...   \n",
       "5329  UNDEF_Y55 gi|696433878|gb|JRIF01000404.1| 1256...   \n",
       "\n",
       "                                               sequence         gene_ID  \\\n",
       "0     ATGAAGTTGAACATTTCTTACCCAGTCAACGGGTCTCAAAAGACCT...  JRIF01000001.1   \n",
       "1     ATGGGTAGAAGAAAAATTGAAATTGAACCTATCAAAGATGATAGAA...  JRIF01000001.1   \n",
       "2     ATGGGAATATTTCGTTGGAACTATCCAGAGAGTTCTGTCCCCGGCG...  JRIF01000001.1   \n",
       "3     ATGTACCAAAATAATGTATTGAATGCTATTTTGGCTTCTGAAAAAT...  JRIF01000001.1   \n",
       "4     ATGAGTGTATTAAGATCTACATGCCTTTTCTTCCCTCCAAGATCCT...  JRIF01000001.1   \n",
       "...                                                 ...             ...   \n",
       "5325  ATGTCACAGTACGGATTTGTAAGAGTTCCTAGAGAGGTAGAAAAGG...  JRIF01000367.1   \n",
       "5326  ATGACAGTAAAGGATAAAAATCAACTAGCTCGATGTGATGCATTGA...  JRIF01000367.1   \n",
       "5327  ATGGACGACATTGAAACAGCCAAGAATCTGACGGTAAAAGCACGTA...  JRIF01000396.1   \n",
       "5328  ATGCCTTATAAAACAGCTATAGATTGCATAGAAGAGTTAGCTACTC...  JRIF01000403.1   \n",
       "5329  ATGTTAAGACAAATCCAAGATAAATTTGGTGGTTCTGTAAAATTAA...  JRIF01000404.1   \n",
       "\n",
       "      count_rare_codon  \n",
       "0                    3  \n",
       "1                    9  \n",
       "2                    9  \n",
       "3                    6  \n",
       "4                    6  \n",
       "...                ...  \n",
       "5325                 3  \n",
       "5326                 3  \n",
       "5327                 7  \n",
       "5328                 2  \n",
       "5329                 0  \n",
       "\n",
       "[5330 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe used\n",
    "genes_y55_df\n",
    "\n",
    "# The rare codon we are interested in\n",
    "rare_cod = \"CGG\"\n",
    "\n",
    "# search for fish, count the number of items in each row\n",
    "rare_occurances = genes_y55_df.sequence.str.split(rare_cod).str.len()\n",
    "\n",
    "# Remove one since if no matches = unsplit = 1 string\n",
    "rare_occurances = rare_occurances-1\n",
    "\n",
    "# Add the count as a column\n",
    "genes_y55_df['count_rare_codon'] = rare_occurances\n",
    "\n",
    "# Display it\n",
    "genes_y55_df\n",
    "\n",
    "# NOTE\n",
    "# Using the following checks that all lines begin with \"ATG\", which they do\n",
    "# sum(genes_y55_df.sequence.str.split(\"^ATG\", regex=True).str.len()-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1721249-04fe-483b-a158-bb6c1655de0f",
   "metadata": {},
   "source": [
    "## Task 6: Search for a motif in all genes\n",
    "Filter the S288C_reference_sequence_R64-4-1_20230830.fsa file for those genes that contain the sequence motif \"CCACACCACACCCAC\".\n",
    "\n",
    "Which chromosome has the most hits?\n",
    "\n",
    "Write the code solution using dataframe manipulation functions. Your code should produce only the chromosome number with the highest motif count.\n",
    "\n",
    "### Setup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e2514f92-7971-4ea1-b4d6-cd1f20555042",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fasta = \"S288C_reference_sequence_R64-4-1_20230830.fsa\"\n",
    "motif = \"CCACACCACACCCAC\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec657ae-3868-484b-a8bf-dd15db570cfc",
   "metadata": {},
   "source": [
    "### From Task 5:\n",
    "\n",
    "Create a Pandas dataframe from the FASTA file\n",
    "\n",
    "*ref: https://stackoverflow.com/questions/19436789/biopython-seqio-to-pandas-dataframe*\n",
    "\n",
    "*ref: https://stackoverflow.com/questions/69696077/i-want-to-parse-sequences-and-sequence-ids-from-a-fasta-file-and-assign-them-to*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a7f0699c-4824-4a17-a301-e46c5681389b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create two empty lists for the sequence IDs and sequences\n",
    "id_list = []\n",
    "sequence_list = []\n",
    "\n",
    "with open(input_fasta) as fasta_file: \n",
    "    # open the file\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        # parse the contents of the FASTA file\n",
    "        # and append the relevant information to the lists\n",
    "        id_list.append(seq_record.description)\n",
    "        sequence_list.append(str(seq_record.seq))\n",
    "\n",
    "# generate a new Pandas dataframe from the two lists\n",
    "# ref: https://www.geeksforgeeks.org/create-a-pandas-dataframe-from-lists/\n",
    "S288C_df = pd.DataFrame(list(zip(id_list, sequence_list)), columns = ['gene_info', 'sequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b2c17eb-70ad-4e3f-9acc-b5f86f5f4266",
   "metadata": {},
   "source": [
    "Also from task 5, but new split pattern to find chromsome number:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ba6d7acb-9889-44be-803b-e415820c856b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to extract the gene IDs from the gene_info columns\n",
    "def extract_gene_id_from_names_S288C(gene_info_col):\n",
    "    # define the function name & set internal argument\n",
    "    return gene_info_col.split(\"] [\")[3]\n",
    "    # split the string into a list using brackets and a space (\"] [\") as the delimiters\n",
    "    # then take the 4th element in the list (Pythonic index 3)\n",
    "    # return that value to outside of the function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bb4bd6e-0f42-4970-9dd0-4d29764c92ea",
   "metadata": {},
   "source": [
    "### New code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9a2717a6-1f0c-424d-a893-7f2ae8f114bb",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'gene_info'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/local/tmp.5733315/ipykernel_1096114/2376792181.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Apply the function and remove the trailing bracket\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mS288C_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'location'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mS288C_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextract_gene_id_from_names_S288C\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremovesuffix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"]\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Split the sequence and remove one, for the same reason as above\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mS288C_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'counts_motif'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mS288C_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmotif\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.10/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6289\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6290\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6291\u001b[0m         ):\n\u001b[1;32m   6292\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6293\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'gene_info'"
     ]
    }
   ],
   "source": [
    "# Apply the function and remove the trailing bracket\n",
    "S288C_df['location'] = S288C_df.gene_info.apply(extract_gene_id_from_names_S288C).str.removesuffix(\"]\")\n",
    "\n",
    "# Split the sequence and remove one, for the same reason as above\n",
    "S288C_df['counts_motif'] = S288C_df.sequence.str.split(motif).str.len()-1\n",
    "\n",
    "# Drop irrelevant columns\n",
    "S288C_df = S288C_df.drop(columns=['gene_info', 'sequence'])\n",
    "\n",
    "# Display only the row with the highest count using the `idxmax` function from pandas\n",
    "S288C_df.loc[S288C_df.counts_motif.idxmax(), :]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
