{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a17e330b-57fc-45db-8d7a-a092500023c6",
   "metadata": {},
   "source": [
    "# Phylogeny Exercise\n",
    "\n",
    "Author: Vi Varga\n",
    "\n",
    "Course: Applied Bioinformatics (BBT045)\n",
    "\n",
    "Due Date: 23.02.2023\n",
    "\n",
    "\n",
    "## Introduction\n",
    "\n",
    "In this exercise you will practice your newly acquired skills in phylogenetic inference and \"tree thinking\", by analyzing the evolutionary history of a gene family. Please note that unlike the tutorial, this exercise should be completed _on your own_. If you have questions, please reach out to me via e-mail (virag.varga@chalmers.se) or message me on Canvas. The assignment should be uploaded to Canvas as a Jupyter Notebook (.ipynb file) by **23.59 on Friday, February 23, 2023**, along with the requested supplementary files. \n",
    "\n",
    "If you do not already have this Jupyter Notebook downloaded (i.e., if you haven't cloned the GitHub repo, [here](https://github.com/V-Varga/BBT045-Phylo-2024/tree/main?tab=readme-ov-file)), you can use `wget` to download the Jupyter Notebook version of this file from [here](PhyloHomework_py.ipynb).\n",
    "\n",
    "In this exercise you are going to follow the phylogenetic analysis workflow discussed during the lecture and tutorial, starting with collecting the data necessary for running the analysis (in the form of homologous protein sequences from different species), through the interpretation of the results (i.e., comparing species trees and gene trees).\n",
    "\n",
    "For this exercise, you will be working with species within a phylum of protists (single-celled eukaryotes) known as the Apicomplexans. Apicomplexans are mostly obligate parasites, meaning that they rely entirely on their host organism to survive. There is debate over the possibility of some species from this phylum being host-associated but non-parasitic, and some recently-discovered Apicomplexans are thought to be free-living, though they are not well-studied (del Campo et al. 2019). \n",
    "\n",
    "![Phylogeny of Apicomplexa](figs/ApicomplexaPhylo__ProtistHandbook.webp)\n",
    "\n",
    "> Source: VotÃ½pka et al. 2016\n",
    "\n",
    "A Newick tree for the clades of interest is provided below (as well as in the Apicomplexa_clades.nwk file): \n",
    "\n",
    "> (Gregarinasia,(Cryptosporidia,((Piroplasmida,Haemosporidia),Coccidia)));\n",
    "\n",
    "When visualized as a cladogram, the tree looks like this: \n",
    "\n",
    "![Phylogenetic tree of key clades in Apicomplexa](figs/Apicomplexa_clades.svg)\n",
    "\n",
    "Below are listed the species you will be working with: \n",
    " - _Porospora gigantea_ \n",
    " - _Plasmodium falciparum_\n",
    " - _Babesia microti_\n",
    " - _Toxoplasma gondii_\n",
    " - _Cryptospodidium parvum_\n",
    "\n",
    "1. Which clade of Apicomplexa do each of the above species belong to? List them below. \n",
    "\n",
    "> - _Porospora gigantea_ : Gregarinasia\n",
    "> - _Plasmodium falciparum_ : Haemosporidia\n",
    "> - _Babesia microti_ : Piroplasmida\n",
    "> - _Toxoplasma gondii_ : Coccidia\n",
    "> - _Cryptospodidium parvum_ : Cryptosporidia\n",
    "\n",
    "You might recognize some of these names! Phylum Apicomplexa contains a number of medically important parasites: \n",
    " - _Cryptosporidium_ species cause Cryptosporidiosis, a water-borne diarrheal disease that often affects immunocompromised people. \n",
    " - _Babesia_ species cause Babesiosis, a tick-borne illness that causes flu-like symptoms. Since the parasite infects and destroys red blood cells, it can also cause anemia. \n",
    " - _Toxoplasma gondii_ causes the disease Toxoplasmosis, which humans contract from food and/or being around pets that are infected. Most infections are asymptomatic, though it can cause flu-like symptoms and serious complications for the fetus if contracted during pregnancy, up to stillbirth or miscarriage. \n",
    " - _Plasmodium falciparum_ is a causative agent of Malaria, a disease which still kills more than half a million people each year. Among the _Plasmodium_ species, _P. falciparum_ causes the most deadly strain of the disease. \n",
    "\n",
    "The Gregarines are comparatively mild parasites - they infect invertebrates, and are rarely fatal to the organisms they infect. \n",
    "\n",
    "2. Why might it be important to study organisms like the Gregarines? (Hint: Take another look at the phylogenetic tree of Apicomplexa.)\n",
    "\n",
    "> Since the Gregarines are related to the others but are not as fatal, helps us find WHAT makes them more fatal. #todo \n",
    "\n",
    "Before moving on to the actual phylogenetic workflow, you should also decide what the outgroup will be. This won't actually come into play until the visualization stage, but it's always good practice to be thinking ahead when following a research pipeline.\n",
    "\n",
    "3. Which group of organisms should be the outgroup? _Why?_\n",
    "\n",
    "> A member of the Gregarines, for example _Porospora gigantea_, since they are closely related to the others but are not pathogenic. #todo too close to the others? \"_What is an outgroup? A species that is related to all organisms in the analysis. But not so closely that it would confuse the phylogenetic signal_\"\n",
    "\n",
    "You may be wondering, \"Why not select an outgroup from elsewhere in Alveolata (the superphylum that Apicompla belongs to)?\" After all, the Gregarines are considered part of Apicomplexa - aren't they too closely related to the rest to be useful in a study like this? \n",
    "\n",
    "In most cases, you might be right! A group of organisms like those in Chromerida, which are also in Alveolata but not considered Apicomplexans, might be a better choice! \n",
    "\n",
    "> (Chromerida,(Gregarinasia,(Cryptosporidia,((Piroplasmida,Haemosporidia),Coccidia))));\n",
    "\n",
    "![Phylogeny of Alveolata, showing Chromerida and Apicomplexa](figs/Apicomplexa_clades_outgroup.svg)\n",
    "\n",
    "In this case, we'll be using the Gregarines because of the gene families presented to you as options - all of these gene families are involved in the \"glideosome,\" which is a key component of Apicomplexan pathogenicity. The glidesome is involved in parasite attachment to, movement along and invasion of host cells. More distantly related species within Alveolata may not have copies of these genes. For example, _C. velia_ doesn't have a copy of the ROM4 gene (Boisard et al. 2022)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0f2c9f2-aef2-41f1-ab23-1f3de1a71b00",
   "metadata": {},
   "source": [
    "## Obtaining & Exploring Data\n",
    "\n",
    "For this exercise, you will be obtaining your own data from the NCBI database. \n",
    "\n",
    "You will start by selecting a gene family to study. Below I've listed a few options: \n",
    " - ROM4\n",
    " - AKMT\n",
    " - MyosinA\n",
    " - DGK1\n",
    "\n",
    "Since _Plasmodium falciparum_ and _Toxoplasma gondii_ have really well-annoted genomes, search for the gene sequence of your choice on the NCBI with, for example: \"Plasmodium falciparum[orgn] ROM4\". Save the header and amino acid sequence to a FASTA file on your local machine. I've provided _recommended_ search combinations below to make things go a bit more smoothly; the members of these protein families do not all include these gene family names in the actual names of the individual genes themselves on the NCBI. If there are multiple results, just go with the first gene hit.\n",
    "\n",
    "Search _recommendations_: \n",
    " - ROM4: _P. falciparum_ ROM4\n",
    " - MyosinA: _P. falciparum_ Myosin-A\n",
    " - AKMT: _T. gondii_ apical complex lysine methyltransferase\n",
    " - DGK1: _T. gondii_  diacylglycerol kinase accessory domain\n",
    "\n",
    "Save the FASTA sequence of your selected reference protein representative of your chosen protein family to a file on your local computer. **This file should be included in your assignment submission.**\n",
    "\n",
    "Next, you'll need to obtain homologous gene sequences from the NCBI database. BLAST your protein sequence on the web to accomplish this, and choose the \"FASTA (complete sequence)\" download option. **This file should be included in your assignment submission.**\n",
    "\n",
    "Note that you can select specific organisms to BLAST against on the NCBI - this significantly reduces the amount of time a search takes, and in our case, makes sure you get only the information you're looking for. There are of course many members of the Apicomplexa that we will not be working with here - no need to sift through all of the gene sequences from those organisms. Include only our 5 species of interest; note that you'll need to select both the A and B versions of _P. gigantea_ individually.\n",
    "\n",
    "4. Open your multiple sequence FASTA and describe some of its characteristics. How many sequences are in the file? How many representatives of each species are there (count the two _P. gigantea_ species as only one)?\n",
    "\n",
    "> There are 75 sequences in the file:\n",
    "> - 31 from Plasmodium falciparum\n",
    "> - 28 from Toxoplasma gondii\n",
    "> - 2 from Babesia microti\n",
    "> - 10 from Porospora gigantea (both)\n",
    "> - 4 from Cryptosporidium parvum\n",
    "\n",
    "5. Describe the following examples with all words that apply from among: homologous, orthologous, paralogous\n",
    "\n",
    " a. Two copies of MyosinA in _B. microti_\n",
    "\n",
    " > Homologous, paralogous #todo Klart?\n",
    "\n",
    " b. Copies of AKMT in _P. falcigarum_ and _T. gondii_\n",
    "\n",
    " > Homologous, ortologous #todo Klart?\n",
    "\n",
    "Before proceeding on to the construction of the MSA, clean your sequences. \n",
    "\n",
    "6. Do you notice any differences in the files? What does this tell you? \n",
    "\n",
    "> The names/headers are standardized with underscores between words instead of spaces or brackets.\n",
    "> The 94 byte count difference comes from that the clean_MSA_seqs has removed 75 closing brackets, 6 commas and 13 colons, which sum to 94. It has therefore not removed any non-standard characters in the sequences. #todo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fa801c81-7987-45f0-a6cb-0b976cfa560e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total sequences:\n",
      "75\n",
      "\n",
      "Plasmodium falciparum:\n",
      "31\n",
      "\n",
      "Toxoplasma gondii:\n",
      "28\n",
      "\n",
      "Babesia microti:\n",
      "2\n",
      "\n",
      "Porospora cf. gigantea:\n",
      "10\n",
      "\n",
      "Cryptosporidium parvum:\n",
      "4\n",
      "\n",
      "Newline, word, and byte counts\n",
      "  555  1039 41077 data/XP_001351627_MSAprep.fasta\n",
      "  555   555 40983 data/XP_001351627_MSAprep_CLEAN.fasta\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Count sequences in fasta file\n",
    "printCount(){\n",
    "    echo -e \"\\n$1:\"\n",
    "    grep -c \"$1\" data/XP_001351627_MSAprep.fasta;}\n",
    "\n",
    "echo \"Total sequences:\"\n",
    "grep -c \"^>\" data/XP_001351627_MSAprep.fasta\n",
    "\n",
    "printCount \"Plasmodium falciparum\"\n",
    "printCount \"Toxoplasma gondii\"\n",
    "printCount \"Babesia microti\"\n",
    "printCount \"Porospora cf. gigantea\"\n",
    "printCount \"Cryptosporidium parvum\"\n",
    "\n",
    "# Clean sequences\n",
    "python clean_MSA_seqs.py data/XP_001351627_MSAprep.fasta\n",
    "\n",
    "echo -e \"\\nNewline, word, and byte counts\"\n",
    "wc data/XP_001351627_MSAprep.fasta\n",
    "wc data/XP_001351627_MSAprep_CLEAN.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0667e76-1f9e-40d3-a543-43b20933751a",
   "metadata": {},
   "source": [
    "## Multiple Sequence Alignment\n",
    "\n",
    "Now that you've familiarized yourself with the data, it's time to create a multiple sequence alignment. \n",
    "\n",
    "7. Write the code you used to create the MSA below. Explain the meanings of the flags, and why you chose to use them.  \n",
    "\n",
    "> I used `localpair` since it is \"_Suitable for a set of locally alignable sequences. Applicable to up to ~200 sequences. A combination with --maxiterate 1000 is recommended (L-INS-i)_\" (ref:[mafft manual](https://mafft.cbrc.jp/alignment/software/manual/manual.html)) and it was the one we used during the tutorial and that worked.\n",
    "> `amino` tells it that the input is a protein FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c7fefcc1-7107-4fea-9017-006732da5b3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: -1 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.520\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   70 / 75\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    70 /74 \n",
      "Reallocating..done. *alloclen = 2991\n",
      "STEP    74 /74 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   70 / 75\n",
      "Segment   1/  1    1-1266\n",
      "STEP 003-039-1  identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    accepted. identical.    accepted. identical.    rejected. rejected. rejected. rejected. rejected. rejected. rejected. accepted. accepted. rejected. rejected. identical.    accepted. rejected. accepted. accepted. rejected. rejected. rejected. accepted. identical.    rejected. identical.    identical.    rejected. rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected.\n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! mafft --localpair --maxiterate 1000 --amino data/XP_001351627_MSAprep_CLEAN.fasta > data/XP_001351627_MSA.fasta\n",
    "# `--localpair --maxiterate 1000` tells MAFFT to use the L-INS-i algorithm\n",
    "# `--amino` tells MAFFT that the input is a protein FASTA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb25eed0-e416-4791-bacc-2c317a00721a",
   "metadata": {},
   "source": [
    "**This MSA file should be included in your assignment submission.**\n",
    "\n",
    "Now open the MSA file in your alignment viewer of choice.\n",
    "\n",
    "8. What patterns do you observe? \n",
    "\n",
    "> When viewing the file with `alen` we observe some regions which are identical between several rows, which indicate that these proteins are similiar to each other. There is one protein which has a lot of insertions, which creates a lot of gaps in the other sequences, the name of this row is: _XP_627451.1_conserved_protein_with_6_transmembrane_domain__Cryptosporidium_parvum_Iowa_II_."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ac3a839-ea9f-405a-85a6-b5affca108b6",
   "metadata": {},
   "source": [
    "## Creating the Tree\n",
    "\n",
    "Now it's time to create a phylogenetic tree. Since you're less pressed for time than you were in-class, use the model finder option of IQ-TREE, if you wish. However, if you'd prefer to use the same tree model as you did during the tutorial, that's fine, too!\n",
    "\n",
    "As you may remember from the in-class tutorial, it has turned out to be rather difficult to run IQ-TREE from within Jupyter Lab, as the program will only be able to access 1 core, even if you've booked two. Therefore, I will instead ask you to run a _batch job_. \n",
    "\n",
    "While we have not discussed them in much detail, batch jobs are what we call the jobs submitted to the backend of an HPC (in this case, Vera) using the SLURM job scheduling system. In order to run IQ-TREE as a batch job, please follow these steps: \n",
    "\n",
    " - Copy the batch script template named `sbatch_iqtree.sh` from the `/cephyr/NOBACKUP/groups/bbt045_2024/Phylogeny/Homework/` directory back to your working directory. \n",
    " - Modify the `sbatch_iqtree.sh` file that you have copied to your working directory, by filling in the necessary information: \n",
    "     - Your Chalmers email (line 7)\n",
    "     - The full path to your working directory (i.e., where this Jupyter Notebook and your homework files are) (line 34)\n",
    "     - The name of your cleaned MSA file (line 41)\n",
    "     - The prefix/file base name you want the IQ-TREE output files to have (line 61)\n",
    " - Submit the batch script with `sbatch sbatch_iqtree.sh` in your _regular_ Vera terminal (i.e., _not_ from within a terminal you open in Jupyter Lab)\n",
    " - Make a record of the job ID that is printed in your terminal when you submit your job. This job ID is necessary in order to be able to trace back any errors/issues, or cancel a running job.\n",
    "\n",
    "In the cell below, fill in the job ID that was printed out when you submitted the job: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd335239-7dee-447a-9c4f-cae389be93fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "sbatch sbatch_iqtree.sh\n",
    "# FILL IN WHAT TEXT IS RETURNED ON THE COMMAND LINE HERE\n",
    "#Submitted batch job 5798794"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33a88a74-5c62-41dc-a165-011e62355572",
   "metadata": {},
   "source": [
    "9. Which tree model did IQ-TREE pick for you, if you used the model finder? Why is the model finding process important?\n",
    "\n",
    "> I didn't use the model finder, but model finding is important since it tests a lot of tree models and find the best one. If you are able to, you should do it instead of skipping it like I did.\n",
    "\n",
    "**The PREFIX.treefile and PREFIX.log files should be included in your assignment submission.**\n",
    "\n",
    "Now open the PREFIX.treefile in a phylogenetic tree visualizer of your choice. \n",
    "\n",
    "10. What does it look like to you? Do you notice anything? \n",
    "\n",
    "> I used the [NCBI Tree Viewer](https://www.ncbi.nlm.nih.gov/projects/treeview/).\n",
    "> #todo\n",
    "\n",
    "Edit the alignment one time, either manually or using a tool. Save the new MSA, and run IQ-TREE on it again, using the model that IQ-TREE previouslsy determined to be the best fit for you. **The new MSA file, PREFIX.treefile and PREFIX.log files should be included in your assignment submission.**\n",
    "\n",
    "11. What method of alignment editing did you choose to use? Does your alignment look different in an alignment viewer? Does the resulting tree look different? How so? If you used a command-line tool, include the code below:\n",
    "\n",
    "> _Fill in your answer!_\n",
    "> #todo\n",
    "> flags used: --gappyout since it \"_Uses an automatic method to decide optimal thresholds, based in the gap percentage count over the whole alignment._\" according to the [TrimAl manual](http://trimal.cgenomics.org/use_of_the_command_line_trimal_v1.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d831b83-8030-4728-9db4-717b5315fb3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Fill in your code here!\n",
    "#Using TrimAl:\n",
    "trimal -in data/XP_001351627_MSA.fasta -out data/XP_001351627_MSA_trim1.fasta -gappyout"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "520d8f42-3687-4d20-a0fa-88881126b2d6",
   "metadata": {},
   "source": [
    "## Comparing the gene tree with the species tree\n",
    "\n",
    "Finally, compare the two gene trees you have visualized with the species tree given to you. (The R script provided to you during the tutorial may help with this, but feel free to do it your own way!) **Include .png/.svg/.jpeg/etc. versions of both of your visualized phylogenetic trees in your assignment submission.**\n",
    "\n",
    "12. How do the gene trees compare to the species tree? Do you notice any interesting patterns? \n",
    "\n",
    "> _Fill in your answer!_\n",
    "\n",
    "13. Is this a good gene family for creating a species tree? Why or why not? \n",
    "\n",
    "> _Fill in your answer!_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97b7ad36-db21-4b56-8c06-04d50621f7ca",
   "metadata": {},
   "source": [
    "## Citations\n",
    "\n",
    "> <div class=\"csl-entry\">Boisard, J., Duvernois-Berthet, E., Duval, L., SchrÃ©vel, J., Guillou, L., Labat, A., le Panse, S., Prensier, G., Ponger, L., &#38; Florent, I. (2022). Marine gregarine genomes reveal the breadth of apicomplexan diversity with a partially conserved glideosome machinery. <i>BMC Genomics</i>, <i>23</i>(1), 1â22. https://doi.org/10.1186/S12864-022-08700-8/FIGURES/5</div> <br>\n",
    "> <div class=\"csl-entry\">del Campo, J., Heger, T. J., RodrÃ­guez-MartÃ­nez, R., Worden, A. Z., Richards, T. A., Massana, R., &#38; Keeling, P. J. (2019). Assessing the diversity and distribution of apicomplexans in host and free-living environments using high-throughput amplicon data and a phylogenetically informed reference framework. <i>Frontiers in Microbiology</i>, <i>10</i>(OCT). https://doi.org/10.3389/fmicb.2019.02373</div> <br>\n",
    "> <div class=\"csl-entry\">TÃ¶pel, M. (2019, November). <i>Webbased Phylogenomic analysis Â· The-Bioinformatics-Group/Teaching Wiki</i>. GitHub. https://github.com/The-Bioinformatics-Group/Teaching/wiki/Webbased-Phylogenomic-analysis</div> <br>\n",
    "> <div class=\"csl-entry\">VotÃ½pka, J., ModrÃ½, D., ObornÃ­k, M., Å lapeta, J., &#38; LukeÅ¡, J. (2016). Apicomplexa. In <i>Handbook of the Protists</i> (pp. 1â58). Springer International Publishing. https://doi.org/10.1007/978-3-319-32669-6_20-1</div> <br>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
